sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java

```diff
@@ -0,0 +1,342 @@
+    }
+
+    PCollection<KV<Integer, Long>> result =
+        p.apply("Int", Create.of(ints)).apply("IntHLL", ApproximateCountDistinct.perKey());
+
+    PAssert.that(result)
+        .containsInAnyOrder(
+            ImmutableList.of(
+                KV.of(0, INTS1_ESTIMATE), KV.of(1, INTS1_ESTIMATE), KV.of(2, INTS1_ESTIMATE)));
+
+    p.run();
+  }
+
+  /** Test Long Globally. */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testStandardTypesPerKeyForLong() {
+
+    List<KV<Integer, Long>> longs = new ArrayList<>();
+
+    for (int i = 0; i < 3; i++) {
+      for (int k : INTS1) {
+        longs.add(KV.of(i, (long) k));
+      }
+    }
+
+    PCollection<KV<Integer, Long>> result =
+        p.apply("Long", Create.of(longs)).apply("LongHLL", ApproximateCountDistinct.perKey());
+
+    PAssert.that(result)
+        .containsInAnyOrder(
+            ImmutableList.of(
+                KV.of(0, INTS1_ESTIMATE), KV.of(1, INTS1_ESTIMATE), KV.of(2, INTS1_ESTIMATE)));
+
+    p.run();
+  }
+
+  /** Test String Globally. */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testStandardTypesPerKeyForStrings() {
+    List<KV<Integer, String>> strings = new ArrayList<>();
+
+    for (int i = 0; i < 3; i++) {
+      for (int k : INTS1) {
+        strings.add(KV.of(i, String.valueOf(k)));
+      }
+    }
+
+    PCollection<KV<Integer, Long>> result =
+        p.apply("Str", Create.of(strings)).apply("StrHLL", ApproximateCountDistinct.perKey());
+
+    PAssert.that(result)
+        .containsInAnyOrder(
+            ImmutableList.of(
+                KV.of(0, INTS1_ESTIMATE), KV.of(1, INTS1_ESTIMATE), KV.of(2, INTS1_ESTIMATE)));
+
+    p.run();
+  }
+
+  /** Test Byte Globally. */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testStandardTypesPerKeyForBytes() {
+
+    List<KV<Integer, byte[]>> bytes = new ArrayList<>();
+
+    for (int i = 0; i < 3; i++) {
+      for (int k : INTS1) {
+        bytes.add(KV.of(i, ByteBuffer.allocate(4).putInt(k).array()));
+      }
+    }
+
+    PCollection<KV<Integer, Long>> result =
+        p.apply("BytesHLL", Create.of(bytes)).apply(ApproximateCountDistinct.perKey());
+
+    PAssert.that(result)
+        .containsInAnyOrder(
+            ImmutableList.of(
+                KV.of(0, INTS1_ESTIMATE), KV.of(1, INTS1_ESTIMATE), KV.of(2, INTS1_ESTIMATE)));
+
+    p.run();
+  }
+
+  /** Test a general object, we will make use of a KV as the object as it already has a coder. */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testObjectTypesGlobal() {
+
+    PCollection<Long> approxResultInteger =
+        p.apply(
+                "Int",
+                Create.of(
+                    INTS1.stream().map(x -> KV.of(x, KV.of(x, x))).collect(Collectors.toList())))
+            .apply(
+                "IntHLL",
+                ApproximateCountDistinct.<KV<Integer, KV<Integer, Integer>>>globally()
+                    .via((KV<Integer, KV<Integer, Integer>> x) -> (long) x.getValue().hashCode()));
```

Comment:
I think the function here should be 'x -> x.hashCode()' without 'getValue'?

Reply:
The value in this case is the KV<int,int> we are using a KV here as a convenient object , so we need the unique from the value which comes from KV<int,KV<int,int>>

Reply:
I see. This tests will pass, but I find it a bit hard to understand. How about we use `x -> x.getValue().getValue()`? I find that simpler to understand. (e.g. KV<x, KV<x, x>> will be identified by x)

Reply:
If we getValue().getValue() then we will get the value of the 2nd KV which is not going to be unique as KV<1,KV<2,3>> is not == to KV<1,KV<3,3>> . If KV is an issue I can create a new POJO just for this test. Using a KV within a KV is just an easy convenience.

Reply:
PS as you have LGTM is it ok to merge as is , or do you need the test changed before ?

Reply:
@robinyqiu are you ok with this being merged without the change to the Test ? Thanx.

Reply:
I am ok with merging it as it is.